### PR TITLE
INSERT/REPLACE complex target column types are validated against source input expressions

### PR DESCRIPTION
**Pull request by zachjsh**
### Description

This change allows for tables that are defined in the catalog to have their complex defined column types validated against source input expressions mapped to them during DML INSERT/REPLACE operations. The catalog dml unit tests have also been greatly simplified, and a lot of duplicate test code simplified or removed.

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java [hunk 1]
```diff
@@ -517,7 +527,17 @@ protected void checkTypeAssignment(
...
+      try {
+        incompatible = !Objects.equals(
+            targetFieldColumnType,
+            ColumnType.leastRestrictiveType(targetFieldColumnType, sourceFieldColumnType)
+        );
+      }
+      catch (Types.IncompatibleTypeException e) {
+        incompatible = true;
+      }
+      if (incompatible) {
```
Comment (kgyrtkirk) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java:540]: * throw a `Types.IncompatibleTypeException` exception based on the result of equals
* you could place these things inside the `catch`
* remove the `incompatible` boolean

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java [hunk 2]
```diff
@@ -475,18 +476,27 @@ private RelDataType validateTargetType(
...
+          relType = RowSignatures.makeComplexType(typeFactory, columnType, sourceField.getType().isNullable());
+        } else {
+          relType = RowSignatures.columnTypeToRelDataType(
+              typeFactory,
+              columnType,
+              // this nullability is ignored for complex types for some reason, hence the check for complex above.
+              sourceField.getType().isNullable()
+          );
+        }
       }
```
Comment (kgyrtkirk) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java:494]: this seems to me that it tries to compute the `RelDataType` for `definedCol` ; can we have method or something for that? maybe that could even make it easier to write more direct tests

File: extensions-core/druid-catalog/src/test/java/org/apache/druid/catalog/sql/CatalogReplaceTest.java [hunk 3]
```diff
@@ -92,7 +92,7 @@ public void buildDatasources()
       createTableMetadata(tableBuilder.build());
     });
     DatasourceFacade catalogMetadata =
-        ((DatasourceTable) resolvedTables.get("foo")).effectiveMetadata().catalogMetadata();
```
Comment (kgyrtkirk) [inline extensions-core/druid-catalog/src/test/java/org/apache/druid/catalog/sql/CatalogReplaceTest.java:95]: note: why not have a value type of `DatasourceTable` for `RESOLVED_TABLES`?

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java (hunk 1 above)
Comment (zachjsh) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java]: Good suggestion, fixed.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java (hunk 2 above)
Comment (zachjsh) [reply sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidSqlValidator.java]: Good call, fixed

File: extensions-core/druid-catalog/src/test/java/org/apache/druid/catalog/sql/CatalogReplaceTest.java (hunk 3 above)
Comment (zachjsh) [reply extensions-core/druid-catalog/src/test/java/org/apache/druid/catalog/sql/CatalogReplaceTest.java:95]: Good question 😅 . Fixed

**Review by kgyrtkirk — APPROVED** [review]

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteCatalogIngestionDmlTest.java [hunk 4]
```diff
@@ -210,11 +237,492 @@ public DruidTable resolveDatasource(
...
+        )
+        .verify();
+  }
+
+  /**
+   * If the segment grain is given in the catalog, and also by PARTITIONED BY, then
+   * the query value is used.
+   */
+  @Test
+  public void testInsertHourGrainWithDayPartitonedByFromQuery()
```
Comment (abhishekrb19) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteCatalogIngestionDmlTest.java:279]:
```suggestion
  public void testUsePartitionedByFromQuery()
```

**Review by abhishekrb19 — APPROVED** [review]
LGTM. Left a few suggestions that can be done in a follow up.

**Pull request merged**